### ISSUE TITLE
`hostDirectives` should auto-resolve double matching

### ISSUE THREAD

**Issue by waterplea** [angular/angular#57846]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

I have been using `hostDirectives` extensively for past several months and `NG0309: Directives can only match an element once.` error is the bane of my existence and a huge burden on Directive Composition API.

I have heard an argument from the Angular team that order of execution matters in case multiple directives have host binding to the same property. I don't believe this reasoning makes any sense when it comes to directives because it's literally impossible to maintain anyway and even if it was — it's a terrible practice to rely on `<div dirA dirB>` behaving differently than `<div dirB dirA>`.

Consider this situation:
All my `buttons` have a host directive that applies some visual rules — `appearance`.
All my `dropdown items` have the same host directive.

Now imagine I have an element, say a `sidebar item`, that can be both a `dropdown item` and a `button` and needs some code from both of those directives. If I add both `button` and `dropdown item` to host directives of `sidebar item` I get `appearance` double matched.

Collapsed sidebar shows items in dropdown:
![image](https://github.com/user-attachments/assets/1d5898e4-02af-41df-8172-10b01ceaa6fa)

Expanded sidebar shows them outside of dropdown as regular buttons:
![image](https://github.com/user-attachments/assets/0ac3c86f-62d5-4831-98ae-dca5327b85ff)

I tried to make this example as simple and transparent as possible to illustrate the issue, please do not take it at face value. In reality this cases are much more complex, but they boil down to this:

**We need to be able to declare host directives freely without having to worry about combining our directives further down the line and finding out some particular combinations cause errors due to nested host directives structure.**

There's an added complexity here because host directive declarations can expose different inputs/outputs, so the real discussion here should be how to handle those best. But the core idea that it shouldn't throw for multiple matches is absolutely clear, unless I'm missing something which I'm sure I'm not but if you have a good reasoning for it — please bring it up.

### Proposed solution

Once a directive is already matched on an element, ignore all subsequent matches like providers do when they are seen multiple times in `providers` arrays of multiple directives on the same element.

### Alternatives considered

None possible

**Comment by waterplea** [comment]
As a library developer, this is rapidly becoming my biggest problem with Angular. Is there a way to at least try/catch it somehow to ignore subsequent matches? At least some crazy nasty hack?

**Comment by splincode** [comment]
+1

**Comment by hakimio** [comment]
@pkozlowski-opensource @alxhub any plans to address this major issue in near future?

**Comment by y-brik** [comment]
+1

**Comment by MillerSvt** [comment]
Currently, it is possible to inject a host directive into a component. However, if we were to allow the instantiation of multiple directives per host, it would raise the question of which directive would be injected.

**Comment by bondarvladislave** [comment]
+1

**Comment by ivanobivan** [comment]
+1

**Comment by mukhammed2021** [comment]
+1

**Comment by waterplea** [comment]
Here's a very clear example of the problem, FYI @crisbeto 

We have a directive for fade using CSS mask:

<img width="353" height="56" alt="Image" src="https://github.com/user-attachments/assets/fd35bd6b-430a-42d3-81ec-0f2fe99faa10" />

The mask kicks in with a **transition** when more text is added than the container can fit or goes away if we scroll it to the end. But we do not want the initial transition to show if the text overflows right away. Therefore we have a host directive that disables initial transition and then puts it back in:

```ts
@Directive({host: {'[style.transition]': '"none"'}})
export class Transitioned {
    private readonly el = inject(ElementRef).nativeElement;

    constructor() {
        afterNextRender(() => this.el.style.setProperty('transition', ''));
    }
}
```

Same undesired transition can happen with other things so we use it elsewhere too. If another directive that also needs this treatment ends up on the same element as fade directive - we get a double match.

Before host directives what we would do is just add multiple selectors to that directive and include it in modules that needed it. Host directives are supposed to make this simpler but they don't.

I'm pretty sure what most people want from host directives is **eagerly instantiated providers with access to host**. I seriously doubt that order of execution is a concern for anybody especially given this is not possible at all at the moment so changing this definitely cannot break anybody.

**Comment by waterplea** [comment]
Here's another good example why double matching **forces** us to use aliasing **all the time**.

1. We have a `Hint` directive in Taiga UI to show hints on elements.
2. We have a `Tooltip` component that is basically question mark icon for showing hints with additional interactions baked in, details of which are irrelevant.
3. The `Tooltip` component has `Hint` as host directive and exposes its inputs so users can pass the actual content, set direction etc.
4. The aliasing of the name is required because otherwise if people import both `Tooltip` and `Hint` directives in the same view, the hint content input, which is the selector for `Hint` directive, would match the directive and it will throw. 

This is just a most boiled down example of a wider problem. Whenever you expose host directive inputs, if they are the same as selectors, which is most of the time, you **have to** use aliasing.

**Comment by waterplea** [comment]
In the PR @crisbeto asked for examples where we might get this issue. Here's my talk doing a deep dive into how we create buttons using several layers of host directives in Taiga UI:
https://www.youtube.com/watch?v=vc5sYiWK7aE

Maybe it can shine some light on how the Directive **Composition** API is used in the wild and how you can easily run into conflicting matches when certain behaviors are isolated into dedicated directives and they are reused across the code base, **as directives should** and as composition implies.

**Comment by intelligentRaji** [comment]
+1

**Comment by splincode** [comment]
+1

**Comment by waterplea** [comment]
I want to re-emphasize that this is a very big problem severely hindering our architectural capabilities. Any plans from the Angular team to address this?

**Comment by imxmv** [comment]
+1

**Comment by crisbeto** [comment]
I'm looking to address this and am interested in some feedback. The context is that when we match a directive, we create a map of all inputs/outputs that are available on it. When we introduce de-duplication, we need some way to consolidate the multiple sources of inputs/outputs. My thinking right now is that we merge the two input/output maps and error if a specific binding is exposed under more than one name.

Right now there are two ways the duplication can occur:
1. A directive appears more than once in the chain of `hostDirectives`.
2. A directive appears both in the template and in the `hostDirectives` chain.

The first case is straightforward, because we can merge the various host directive configs together. The second one is a bit trickier, because it's unclear which inputs/outputs can be exposed. I see a few solutions, but I don't think that any of them are ideal:

1. We treat the template-matched directive as a "host directive" that exposes all of its inputs/outputs under their original names. This approach pretty much guarantees a conflict, assuming that at least one input/output is exposed by a "real" host directive.
2. We ignore the template-matched instance of the directive and we only look at the host directive instances. I'm not sure if this is a good idea, because the consumer in the template might be expecting to use the directive's original inputs/outputs, given that they explicitly wrote out its selector. I suspect this will also complicate template type checking.
3. We throw an error along the lines of "Directive can't match both in the template and as a host directive". This one makes the most sense IMO, but it doesn't fully remove the error that this issue is asking for.
4. We ignore the host directives and only apply the template-matched one. This one is more or less the inverse of option two.

I'm leaning toward the third solution, but I'm not sure if this actually addresses the problems you were having.

cc @waterplea @splincode @MillerSvt

**Comment by splincode** [comment]
Well, as I understand it, the current model is too fragile for deep composition, even though the documentation does not explicitly say so. Let me break down whether I understood the current problem correctly.

#### Case 1. Take this directive:

```ts
@Directive({
    standalone: true,
})
export class TooltipDirective {
    delay = input(0);
    text = input('');
    closed = output<void>();
}
```

Now two composed directives:

```ts
@Directive({
    standalone: true,
    hostDirectives: [
        {
            directive: TooltipDirective,
            inputs: ['delay: openDelay'],
            outputs: ['closed: tooltipClosed'],
        },
    ],
})
export class MenuBehavior {}

@Directive({
    standalone: true,
    hostDirectives: [
        {
            directive: TooltipDirective,
            inputs: ['text: hint'],
        },
    ],
})
export class HoverBehavior {}
```

And then:

```ts
@Component({
    standalone: true,
    selector: 'app-button',
    template: '',
    hostDirectives: [MenuBehavior, HoverBehavior],
})
export class AppButton {}
```

If Angular deduplicates `TooltipDirective`, it has to decide:

- should `openDelay` be preserved?
- should `hint` be preserved?
- should `tooltipClosed` be preserved?
- and if two chains expose the same input under different names, should that be treated as an error or not?


#### Case 2.

Angular ignores the selector of directives applied through `hostDirectives` and instantiates those host directives separately on the host element. At the same time, if the user explicitly applies the directive in the template, Angular would normally be expected to treat it through its regular public API.

```ts
@Directive({
    selector: '[tooltip]',
    standalone: true,
})
export class TooltipDirective {
    delay = input(0);
    text = input('');
    closed = output<void>();
}

@Component({
    standalone: true,
    selector: 'app-card',
    template: '',
    hostDirectives: [
        {
            directive: TooltipDirective,
            inputs: ['delay: tooltipDelay'],
        },
    ],
})
export class AppCard {}
```

```html
<app-card tooltip [delay]="100" [tooltipDelay]="200" />
```

Now the question is:

- is this a single `TooltipDirective` instance?
- which bindings are actually available on it: `delay`, `tooltipDelay`, or both?
- if both point to the same internal input, which one wins?


The problem here is clear, and it makes sense why this is not as straightforward as it may seem. This is exactly why the second case is much more problematic than the first one.

----

@crisbeto Now let’s go through your proposed options to make sure I understood your ideas correctly.

#### Solution 1

Treat the template-matched directive as a host directive that exports all of its inputs/outputs under their original names. In other words, if the directive is explicitly written in the template, Angular assumes that it automatically exposes its entire regular public API.

```ts
@Directive({
    selector: '[tooltip]',
    standalone: true,
})
export class TooltipDirective {
    delay = input(0);
    text = input('');
    closed = output<void>();
}

@Component({
    standalone: true,
    selector: 'app-card',
    template: '',
    hostDirectives: [
        {
            directive: TooltipDirective,
            inputs: ['delay: tooltipDelay'],
            outputs: ['closed: tooltipClosed'],
        },
    ],
})
export class AppCard {}
```

```html
<app-card
    tooltip
    [delay]="100"
    [tooltipDelay]="200"
    (closed)="a()"
    (tooltipClosed)="b()"
/>
```

The template-matched version says:

- `delay` is available
- `text` is available
- `closed` is available

The host-directive version says:

- `tooltipDelay` is available
- `tooltipClosed` is available

##### Pros

- A user who explicitly writes tooltip in the template gets the familiar API. 
- Template type checking looks logical.

##### Cons

- Conflicts are almost guaranteed. For example, if `hostDirectives` already exposes    `delay` or `closed`, we end up with overlapping names.

#### Solution 2

Ignore the template-matched instance and only consider the host directive. In other words, if the directive comes both from the template and from hostDirectives, Angular says: only the host-directive configuration is treated as the real one.

Same classes and:

```html
<app-card tooltip [delay]="100" [tooltipDelay]="200" />
```

But Angular only allows what is exposed through hostDirectives:

- `tooltipDelay` — yes
- `delay` — no

##### Pros

- The model is simpler. It becomes possible to merge only the host-directive configuration consistently.

##### Cons

- This breaks user expectations. The user wrote tooltip in the template and would naturally expect this to work:

```html
<app-card tooltip [delay]="100" />
```

But Angular would say: no, only `tooltipDelay` can be used.

#### Solution 3

Throw an error: a directive cannot match both in the template and as a host directive at the same time. In other words:

- if the duplicate appears only within the hostDirectives chain, Angular can deduplicate and merge it
- if the duplicate appears because the directive exists both in the template and in hostDirectives, Angular throws a separate, clear error

```ts
@Component({
    standalone: true,
    selector: 'app-card',
    template: '',
    hostDirectives: [
        {
            directive: TooltipDirective,
            inputs: ['delay: tooltipDelay'],
        },
    ],
})
export class AppCard {}
```

If I try do this:

```html
<app-card tooltip />
```

Then Angular would say something like: `TooltipDirective cannot match both from the template and from hostDirectives`

##### Core idea

This is a compromise:

- host-directive vs host-directive => can be deduplicated
- template vs host-directive => forbidden as ambiguous

In practice, this does seem like the safest option.

##### Pros

- The semantics are clear.
- There is no need to decide which one wins: the template or the host directive.
- There is no need to merge two different public APIs.
- It is simpler for the compiler and for template type checking.

##### Cons

- This does not eliminate all errors, it only changes their form.
- Our issue is specifically asking to remove the pain caused by repeated matching in general, whereas with this approach some scenarios would still remain errors.

#### Solution 4

Ignore host directives and only apply the template-matched directive. In other words, if the directive is explicitly present in the template, Angular treats that instance as the primary one and simply discards the host-directive configuration for that directive.

```html
<app-card tooltip [delay]="100" [tooltipDelay]="200" />
```

Angular would keep only the regular template API:

- `delay` — yes
- `text` — yes
- `closed` — yes
- `tooltipDelay` — no

##### Core idea

The component says, "I want to expose tooltipDelay" but the template says, "No, I attached tooltip myself, so give me delay".

##### Pros

- The template behavior is predictable for the HTML author.
- If the directive is explicitly written in the template, that is treated as the higher-priority source.

##### Cons

- The API that the component intended to expose through hostDirectives breaks silently.
- The composed component loses part of its contract.
- The behavior becomes non-obvious for the component author.

----

What do you mean by merging maps? I hope I understand the idea correctly. Below I will describe a few examples.

#### The simplest example. 

There is a single directive:

```ts
@Directive({
    standalone: true,
})
export class AppearanceDirective {
    color = input('primary');
    size = input('m');
}
```

Now there are two wrappers around it.

The first one:

```ts
@Directive({
    standalone: true,
    hostDirectives: [
        {
            directive: AppearanceDirective,
            inputs: ['color: buttonColor'],
        },
    ],
})
export class ButtonBehavior {}
```

```ts
@Directive({
    standalone: true,
    hostDirectives: [
        {
            directive: AppearanceDirective,
            inputs: ['size: itemSize'],
        },
    ],
})
export class ItemBehavior {}
```

And then the component:

```ts
@Component({
    standalone: true,
    selector: 'app-demo',
    template: '',
    hostDirectives: [ButtonBehavior, ItemBehavior],
})
export class DemoComponent {}
```

What Angular sees internally (I guess). Not literally this code, but conceptually something like this:

From `ButtonBehavior`:

```ts
inputsMapA = {
    buttonColor: 'color',
};
```

From `ItemBehavior`:

```ts
inputsMapB = {
    itemSize: 'size',
};
```

After merging:

```ts
mergedInputsMap = {
    buttonColor: 'color',
    itemSize: 'size',
};
```

This is a good merge.

Why is it good?

- the public names are different
- they point to different actual inputs
- the resulting component API is clear

```html
<app-demo [buttonColor]="'red'" [itemSize]="'l'" />
```

#### Example where the merge becomes questionable

Now both paths expose the same underlying input, but under different public names.

The first wrapper:

```ts
@Directive({
    standalone: true,
    hostDirectives: [
        {
            directive: AppearanceDirective,
            inputs: ['color: buttonColor'],
        },
    ],
})
export class ButtonBehavior {}
```

```ts
@Directive({
    standalone: true,
    hostDirectives: [
        {
            directive: AppearanceDirective,
            inputs: ['color: itemColor'],
        },
    ],
})
export class ItemBehavior {}
```

The component:

```ts
@Component({
    standalone: true,
    selector: 'app-demo',
    template: '',
    hostDirectives: [ButtonBehavior, ItemBehavior],
})
export class DemoComponent {}
```

The resulting maps are:

```ts
inputsMapA = {
    buttonColor: 'color',
};

inputsMapB = {
    itemColor: 'color',
};
```

If we merge them mechanically:

```ts
mergedInputsMap = {
    buttonColor: 'color',
    itemColor: 'color',
};
```

In other words, from the outside both of these would be possible:

```html
<app-demo [buttonColor]="'red'" />
<app-demo [itemColor]="'red'" />
```

But what if someone writes both?

```html
<app-demo [buttonColor]="'red'" [itemColor]="'blue'" />
```

In that case, both bindings point to the same `AppearanceDirective.color`. Which one is supposed to win? Am I understanding correctly that this should result in a compilation error?

#### The most obvious conflict

Now an even simpler case: two paths want to expose the same public name, but for different underlying inputs.

The first wrapper:

```ts
@Directive({
    standalone: true,
    hostDirectives: [
        {
            directive: AppearanceDirective,
            inputs: ['color: appearance'],
        },
    ],
})
export class ButtonBehavior {}
```

The second wrapper:

```ts
@Directive({
    standalone: true,
    hostDirectives: [
        {
            directive: AppearanceDirective,
            inputs: ['size: appearance'],
        },
    ],
})
export class ItemBehavior {}
```

The maps are:

```ts
inputsMapA = {
    appearance: 'color',
};

inputsMapB = {
    appearance: 'size',
};
```

After merging, we get an absurd result:

```ts
mergedInputsMap = {
    appearance: ???,
};
```

because it is unclear what this means:

```html
<app-demo [appearance]="..." />
```

Is it color or size? A case like this should fail unconditionally.

#### The safest case

If both chains expose exactly the same thing in exactly the same way, then the merge can be done safely.

The first one:

```ts
inputsMapA = {
    buttonColor: 'color',
};
```

The second one:

```ts
inputsMapB = {
    buttonColor: 'color',
};
```

Merge:

```ts
mergedInputsMap = {
    buttonColor: 'color',
};
```

In other words, the duplicate exists, but it is fully identical. That kind of merge can simply be deduplicated.

#### The same applies to outputs

For example:

```ts
@Directive({
    standalone: true,
})
export class PopupDirective {
    closed = output<void>();
}
```

One chain:

```ts
outputsMapA = {
    popupClosed: 'closed',
};
```

Another chain:

```ts
outputsMapB = {
    menuClosed: 'closed',
};
```

After merging:

```ts
mergedOutputsMap = {
    popupClosed: 'closed',
    menuClosed: 'closed',
};
```

And again the question is: can the same output, `closed`, be listened to under two different names?

```html
<app-demo (popupClosed)="a()" (menuClosed)="b()" />
```

Formally, it is possible to define semantics for this. But the API becomes strange and ambiguous.

----

These are all the examples I wanted to go through to make sure I understand what you mean. I hope I got the idea right.

This may have been a bit lengthy and more example-heavy than necessary.

**Comment by splincode** [comment]
So, to summarize what I understood and what I would propose:

### merge API

1) Always deduplicate the runtime instance by directive class. If `TooltipDirective` reaches the element through two different paths, there should still be only one `TooltipDirective` instance on that element.

2) Merge not the directives themselves, but the binding descriptors.
 
3) Introduce clear conflict rules.

#### Inputs

-  Safe

```ts
buttonColor => color
itemSize    => size
```

Different public names point to different underlying bindings.

- Also safe

```ts
buttonColor => color
buttonColor => color
```

This is just deduplication.

- Error

```ts
appearance => color
appearance => size
```

The same public name points to different underlying bindings.

- Ambiguous, but still solvable

```ts
buttonColor => color
itemColor   => color
```

Two different public names point to the same underlying input.

My preferred handling for this ambiguous case would be:

```ts
@Directive({standalone: true})
export class AppearanceDirective {
    color = input('primary');
}

inputsMapA = {buttonColor: 'color'};
inputsMapB = {itemColor: 'color'};
```

The resulting API may be allowed:

```html
<app-demo [buttonColor]="'red'" />
<app-demo [itemColor]="'red'" />
```

But this should be an error:

```html
<app-demo [buttonColor]="'red'" [itemColor]="'blue'" />
```

because both bindings write to the same underlying input.

##### Why this seems reasonable

- it does not discard useful aliases
- it preserves an ergonomic API
- the conflict is detected at the point of actual use, rather than abstractly while building the graph

#### Outputs

For outputs, I would be more permissive: multiple aliases for the same underlying output could be allowed.

```ts
outputsMapA = {popupClosed: 'closed'};
outputsMapB = {menuClosed: 'closed'};
```

This does not break the directive’s state. It is simply two names for the same event stream.

However, I would add a warning-level diagnostic if the template subscribes to both:

```html
<app-demo (popupClosed)="a()" (menuClosed)="b()" />
```

This can work, but at that point the API already becomes somewhat strange.

### template + host duplicate

@crisbeto Here I would suggest a hybrid of your solutions 1 and 3.

#### Rule

If a directive is matched both in the template and through hostDirectives, then:

- the original names from the template are always available
- host aliases are allowed only if they do not conflict at the public-name level
- if the same underlying input is available both under its original name and under an alias, that is allowed
- however, if the user binds both of them at the same time, that should be a compile-time error

```ts
@Directive({
    selector: '[tooltip]',
    standalone: true,
})
export class TooltipDirective {
    delay = input(0);
}

@Component({
    standalone: true,
    selector: 'app-card',
    template: '',
    hostDirectives: [
        {
            directive: TooltipDirective,
            inputs: ['delay: tooltipDelay'],
        },
    ],
})
export class AppCard {}
```

This is fine:

```html
<app-card tooltip [delay]="100" />
<app-card tooltip [tooltipDelay]="100" />
```

But this should already be an error:

```html
<app-card tooltip [delay]="100" [tooltipDelay]="200"></app-card>
```

#### Why I like this option

It preserves:

- the template author’s expectations
- the usefulness of aliases exposed through the host API
- predictable compile-time diagnostics

And it is better than simply saying always error, because some of these cases are genuinely non-conflicting.

-----

To summarize, duplication by itself should not be considered an error. The error should be an ambiguous public API.

**Comment by crisbeto** [comment]
What I mean by "merging maps" is that if you have one host directive definition that has `inputs: ['foo']` and another one that has `input: ['bar']`, they get treated as `inputs: ['foo', 'bar']`. An example of a conflict would be if the first directive had `inputs: ['foo']` while the second had `input: ['foo: alias']`. We can't expose the same input under more than one name, because it might have different values and it wouldn't be clear which value we should take.

**Comment by splincode** [comment]
Would it make sense to consider a hybrid of options 1 and 3, where the template-matched directive keeps its original public API, while host directive bindings are merged only if they don't expose the same real input/output under a different public name, as I described in my comment above?

**Comment by waterplea** [comment]
First of all, throwing in runtime without any ways for us to handle the issue is a bad idea. 

Currently we are forced to alias all the time, I explained why. The merging with throwing if the names are different would make version before this change and after incompatible which is also bad. It might be ok when we have multiple hostDirectives conflicting, that can be handled by lib authors by just using the same alias in them. But if one of those directives is template applied, I think its naming should take priority instead of throwing.

**Comment by crisbeto** [comment]
I can see an argument for the following approach:
1. If there are multiple host-directive-only matches, we merge them and throw if two aliases point to the same input/output. I think that we need to throw, because reconciling multiple sources for the same input can get messy.
2. If the directive matches in the template and through host directives, we expose the entire directive and disregard any of the host directive aliases.

My thinking here is that we can consider host directives as `Partial<SomeDirective>` whereas a template-matched instance is the full `SomeDirective`.

**Comment by waterplea** [comment]
Right, if the directive is in the template we  already have all inputs/outputs and we can indeed disregard host directives. I like your reasoning and mental model 👍 it means that libraries made with older Angular and aliasing to avoid double matching can be safely used in modern version without aliasing.

**Comment by crisbeto** [comment]
There's a fix at #67996.

**Comment by waterplea** [comment]
Thank you very much, @crisbeto! Is it coming to v21 or the upcoming v22?

**Comment by crisbeto** [comment]
It's targeted for v22.

**Comment by waterplea** [comment]
Would be super great to get it into 21 as well, otherwise all libraries that need this would force their users apps to update, but that's a big ask in the enterprise world.

**Comment by crisbeto** [comment]
We generally don't backport new features.
